src/middlewared/middlewared/plugins/smb.py

```diff
@@ -110,3 +112,286 @@ class Config:
         await self._update_service(old, new)
 
         return await self.config()
+
+
+class SharingCIFSService(CRUDService):
+    class Config:
+        namespace = 'sharing.cifs'
+        datastore = 'sharing.cifs_share'
+        datastore_prefix = 'cifs_'
+        datastore_extend = 'sharing.cifs.extend'
+
+    @accepts(Dict(
+        'sharingcifs_create',
+        Str('path'),
+        Bool('home'),
+        Str('name'),
+        Str('comment'),
+        Bool('ro'),
+        Bool('browsable'),
+        Bool('recyclebin'),
+        Bool('showhiddenfiles'),
+        Bool('guestok'),
+        Bool('guestonly'),
+        Bool('abe'),
+        List('hostsallow', items=[IPAddr('ip', cidr=True)]),
+        List('hostsdeny', items=[IPAddr('ip', cidr=True)]),
+        List('vfsobjects'),
+        Int('storage_task'),
+        Str('auxsmbconf'),
+        Bool('default_permissions'),
+        register=True
+    ))
+    async def do_create(self, data):
+        verrors = ValidationErrors()
+        path = data['path']
+        default_perms = data.pop('default_permissions', True)
+
+        await self.clean(data, 'sharingcifs_create', verrors)
+        await self.validate(data, 'sharingcifs_create', verrors)
+
+        if path and not os.path.exists(path):
```

Comment:
I just thought of something, this does not enforce path has to live within a volume, which would allow you to create a path in /usr/local/samba. Django forbid us to do that, the API does not.

We gotta use `check_path_resides_within_volume` async validator. We may have forgotten that in other plugins.